### PR TITLE
Upstream max conn upgrade

### PR DESCRIPTION
**Pull request by SolidWallOfCode**
This is a refresh of the upstream connection tracking which supports throttling per upstream. The changes are

* Much less locking - the table of upstream records is locked only once and a pointer kept for the rest of the session life.
* Atomic values for tracking.
* Maximum connection count per record.
* Alert throttling - at most every 60 seconds per record. Actions between alerts are tracked and provided in the next alert method.
* Correct behavior - the previous version leaked connection counts and was therefore too unreliable to use in production. This version has been tested under severe circumstances in Oath production and proven reliable and effective.
* Prevent process lock - due to the removal of the congestion support in general, use of this feature would frequently cause a `ET_NET` thread to spin and stop processing transactions.
* Better naming for configuration variables.
* Separate grouping from session sharing.
* More configuration options.
* Store the host name for better monitoring.

Depends on #3800, #3818, #3817.

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
[approve ci centos]

**Comment by SolidWallOfCode** [comment]
[approve ci centos]

**Comment by vmamidi** [comment]
@SolidWallOfCode Are we seeing any problems with the existing connection tracking code and how is this change addressing that problem?

**Comment by SolidWallOfCode** [comment]
In our production experience, the existing upstream connection tracking just doesn't work. It gets the counts wrong and generally high. Under low load and happy origins it seems to work but under load and/or stressed origins connection counts are lost and throttling starts inappropriately. For instance, with a setting of 3000, we'd get throttling when `ss -s` claimed 538 total sockets open. I spent a long time trying to track down where there was happening, but eventually it was faster to do a full refresh and get it correct.

In addition it had performance issues, because every access was a hash table lookup with a lock. This version does one lookup/lock per outbound session.

After deploying this in production, we went from regular reports of excessive throttling to surviving multiple near meltdowns with full recovery, post incident spots checks showing the connection counts were still accurate.

Basically, this solves a chronic production problem for us, along with providing better instrumentation (e.g., tracking the high water mark for adjusting the throttle values).

File: proxy/http/HttpServerSession.cc [hunk 1]
```diff
@@ -117,30 +118,35 @@ HttpServerSession::do_io_shutdown(ShutdownHowTo_t howto)
...
-      ip_port_text_buffer addrbuf;
-      Debug("http_ss", "[%" PRId64 "] connection closed, ip: %s, count: %u", con_id,
-            ats_ip_nptop(&get_server_ip().sa, addrbuf, sizeof(addrbuf)),
-            connection_count->getCount(get_server_ip(), hostname_hash, sharing_match));
+  // Update upstream connection tracking data if present.
+  if (conn_track_group) {
+    if (conn_track_group->_count >= 0) {
+      auto n = (conn_track_group->_count)--;
+      if (debug_p)
+        w.print(" conn track group ({},{},{:s}) count {}", conn_track_group->_addr, conn_track_group->_fqdn_hash,
```
Comment (bryancall) [inline proxy/http/HttpServerSession.cc]: Doing this can't print the filename and line number in the code.

File: proxy/http/HttpConnectionCount.h [hunk 2]
```diff
@@ -21,237 +21,241 @@
...
     }
+  };
 
-    ConnAddr(const IpEndpoint &addr, const char *hostname, TSServerSessionSharingMatchType match_type)
-      : _addr(addr), _match_type(match_type)
-    {
-      CryptoContext().hash_immediate(_hostname_hash, static_cast<const void *>(hostname), strlen(hostname));
-    }
+  /// Container type for the connection groups.
+  using HashTable = TSHashTable<HashDescriptor>;
```
Comment (bryancall) [inline proxy/http/HttpConnectionCount.h]: I would remove this, since it is only used once.

File: proxy/http/HttpTransact.h [hunk 3]
```diff
@@ -755,8 +759,10 @@ class HttpTransact
     bool is_websocket        = false;
     bool did_upgrade_succeed = false;
 
-    // Some queue info
-    bool origin_request_queued = false;
+    // Track if this SM has reserved (incremented) the outbound connection tracking data.
```
Comment (bryancall) [inline proxy/http/HttpTransact.h]: Seems like this should be part of the class that keeps track of state and has methods around some of the logic instead of having the state variables exposed in `HttpTransact::State` and having the logic in `HttpSM::do_http_server_open`

File: proxy/http/HttpSM.cc [hunk 4]
```diff
@@ -4860,56 +4868,99 @@ HttpSM::do_http_server_open(bool raw)
...
+    if (ccount > t_state.txn_conf->origin_max_connections) {
+      --ct_info._count; // cancel our reservation
+      t_state.outbound_conn_reserved = false;
+
+      ink_assert(pending_action == nullptr); // in case of reschedule must not have already pending.
+
+      // If the queue is disabled, reschedule.
+      if (t_state.txn_conf->origin_max_connections_queue < 0) {
+        ++ct_info._rescheduled;
         pending_action = eventProcessor.schedule_in(this, HRTIME_MSECONDS(100));
```
Comment (bryancall) [inline proxy/http/HttpSM.cc]: This is a bad default and I wonder if the option should even exist.  Allowing an unlimited number state machines to be rescheduled every 100ms with no upper bounds on the number of times to be rescheduled is not a good idea.

File: proxy/http/HttpSM.cc [hunk 5]
```diff
@@ -4860,56 +4868,99 @@ HttpSM::do_http_server_open(bool raw)
...
         // if there is space in the queue
-        if (waiting_connections->getCount(t_state.current.server->dst_addr, hostname_hash,
-                                          (TSServerSessionSharingMatchType)t_state.txn_conf->server_session_sharing_match) <
-            t_state.txn_conf->origin_max_connections_queue) {
-          t_state.origin_request_queued = true;
-          Debug("http", "[%" PRId64 "] queued for this host: %s", sm_id,
-                ats_ip_ntop(&t_state.current.server->dst_addr.sa, addrbuf, sizeof(addrbuf)));
-          waiting_connections->incrementCount(t_state.current.server->dst_addr, hostname_hash,
-                                              (TSServerSessionSharingMatchType)t_state.txn_conf->server_session_sharing_match, 1);
+        if (wcount < t_state.txn_conf->origin_max_connections_queue) {
```
Comment (bryancall) [inline proxy/http/HttpSM.cc]: I think we should remove the queue and return and error response to the client instead of rescheduling.

File: proxy/http/HttpSM.cc (hunk 4 above)
Comment (SolidWallOfCode) [reply proxy/http/HttpSM.cc]: That's how the previous version worked, I was trying to minimize functional changes.

File: proxy/http/HttpSM.cc (hunk 5 above)
Comment (SolidWallOfCode) [reply proxy/http/HttpSM.cc]: That's a breaking change which I think should be discussed on the dev list. I tracked it down and the rescheduling behavior dates back to before the initial commit to git. Thomas Jackson added the max queue length option to ameliorate the problem.

File: proxy/http/HttpTransact.h (hunk 3 above)
Comment (SolidWallOfCode) [reply proxy/http/HttpTransact.h]: Isn't `HttpTransact::State` the class that keeps track of state for the HTTP state machine?

**Comment by SolidWallOfCode** [comment]
I changed the default of `proxy.config.http.origin_max_connections_queue` to 0 which disables the flying dutchman queue by default.

File: proxy/http/HttpSM.cc (hunk 4 above)
Comment (bryancall) [reply proxy/http/HttpSM.cc]: Since it is rewritten, the design should be looked at instead of just copying the existing functionality.

File: proxy/http/HttpTransact.h (hunk 3 above)
Comment (bryancall) [reply proxy/http/HttpTransact.h]: The new class would be a member of `HttpTransact::State`.  The new implementation, much like the pervious, has part of it scatters all over the place.

File: lib/ts/apidefs.h.in [hunk 6]
```diff
@@ -578,6 +579,15 @@ typedef enum {
   TS_SERVER_SESSION_SHARING_POOL_GLOBAL,
   TS_SERVER_SESSION_SHARING_POOL_THREAD,
 } TSServerSessionSharingPoolType;
+
+/// Values for per server outbound connection tracking group definition.
+/// See proxy.config.http.per_server.match
+typedef enum {
```
Comment (SolidWallOfCode) [inline lib/ts/apidefs.h.in]: This is used for the per transaction override.

File: lib/ts/bwf_std_format.h [hunk 7]
```diff
@@ -25,19 +25,17 @@
 
 #include <atomic>
 #include <string_view>
+#include <ts/BufferWriter.h>
 
-namespace std
+namespace ts
```
Comment (SolidWallOfCode) [inline lib/ts/bwf_std_format.h]: This seems to work better than being in `std`.

File: mgmt/MgmtDefs.h [hunk 8]
```diff
@@ -61,6 +63,35 @@ typedef enum {
  */
 typedef void *(*MgmtCallback)(void *opaque_cb_data, char *data_raw, int data_len);
 
+//-------------------------------------------------------------------------
```
Comment (SolidWallOfCode) [inline mgmt/MgmtDefs.h]: Added this so the contributions to the HTTP config structures could be easier to work with types compared to `MgmtInt`. It also enables setting the group match by name instead of by enumeration value.

File: mgmt/RecordsConfig.cc [hunk 9]
```diff
@@ -371,7 +371,7 @@ static const RecordElement RecordsConfig[] =
   ,
   {RECT_CONFIG, "proxy.config.http.server_session_sharing.match", RECD_STRING, "both", RECU_DYNAMIC, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
   ,
-  {RECT_CONFIG, "proxy.config.http.server_session_sharing.pool", RECD_STRING, "thread", RECU_RESTART_TS, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
+  {RECT_CONFIG, "proxy.config.http.server_session_sharing.pool", RECD_STRING, "thread", RECU_DYNAMIC, RR_NULL, RECC_NULL, nullptr, RECA_NULL}
```
Comment (SolidWallOfCode) [inline mgmt/RecordsConfig.cc]: Why did I do this? Need to revert.

File: proxy/http/HttpConfig.cc [hunk 10]
```diff
@@ -929,9 +929,7 @@ HttpConfig::startup()
   HttpEstablishStaticConfigLongLong(c.server_max_connections, "proxy.config.http.server_max_connections");
   HttpEstablishStaticConfigLongLong(c.max_websocket_connections, "proxy.config.http.websocket.max_number_of_connections");
   HttpEstablishStaticConfigLongLong(c.oride.server_tcp_init_cwnd, "proxy.config.http.server_tcp_init_cwnd");
-  HttpEstablishStaticConfigLongLong(c.oride.origin_max_connections, "proxy.config.http.origin_max_connections");
-  HttpEstablishStaticConfigLongLong(c.oride.origin_max_connections_queue, "proxy.config.http.origin_max_connections_queue");
-  HttpEstablishStaticConfigLongLong(c.origin_min_keep_alive_connections, "proxy.config.http.origin_min_keep_alive_connections");
+  HttpEstablishStaticConfigLongLong(c.origin_min_keep_alive_connections, "proxy.config.http.per_server.min_keep_alive");
```
Comment (SolidWallOfCode) [inline proxy/http/HttpConfig.cc]: Although renamed in this PR, this value doesn't really interact with the others. The config update logic for the main values has been moved to the implementation file, rather than being here.

File: proxy/http/HttpConfig.cc [hunk 11]
```diff
@@ -1171,6 +1169,8 @@ HttpConfig::startup()
   HttpEstablishStaticConfigLongLong(c.oride.number_of_redirections, "proxy.config.http.number_of_redirections");
   HttpEstablishStaticConfigLongLong(c.post_copy_size, "proxy.config.http.post_copy_size");
 
+  OutboundConnTrack::global_init(&c.outbound_conntrack, &c.oride.outbound_conntrack);
```
Comment (SolidWallOfCode) [inline proxy/http/HttpConfig.cc]: This sets up the config update callbacks and does the initial load.

File: proxy/http/HttpConfig.h [hunk 12]
```diff
@@ -736,6 +733,8 @@ struct OverridableHttpConfigParams {
   MgmtInt default_buffer_water_mark;
   MgmtInt slow_log_threshold;
 
+  OutboundConnTrack::TxnConfig outbound_conntrack;
```
Comment (SolidWallOfCode) [inline proxy/http/HttpConfig.h]: This is something I've discussed with Leif and others, to simplify the HTTP configuration structures. Rather than declaring the configuration for a subsystem here, it's declared in the subsystem and included via member variable here. Updates are then more localized. More importantly, when it is necessary to pass this configuration to the subsystem for a specific transaction, just this structure can be passed and therefore the subsystem doesn't have to include all of the HTTP configuration headers. This is simpler and removes one major cause of circular dependencies. Now the HTTP configuration depends on the subsystem, but not the reverse.

File: proxy/http/HttpConnectionCount.cc [hunk 13]
```diff
@@ -21,51 +21,321 @@
...
+#include <ts/bwf_std_format.h>
 
-ConnectionCount ConnectionCount::_connectionCount;
-ConnectionCountQueue ConnectionCountQueue::_connectionCount;
+OutboundConnTrack::Imp OutboundConnTrack::_imp;
+// const std::chrono::seconds OutboundConnTrack::Group::ALERT_DELAY{60};
+
+OutboundConnTrack::GlobalConfig *OutboundConnTrack::_global_config{nullptr};
+
+const MgmtConverter OutboundConnTrack::MAX_CONV{
```
Comment (SolidWallOfCode) [inline proxy/http/HttpConnectionCount.cc]: These convert between TS API call values and the implementation values.

File: proxy/http/HttpConnectionCount.cc [hunk 14]
```diff
@@ -21,51 +21,321 @@
...
+    return t < 0 || t > OutboundConnTrack::MATCH_BOTH ? "Invalid"sv : OutboundConnTrack::MATCH_TYPE_NAME[t];
+  },
+  [](void *data, std::string_view src) -> void {
+    OutboundConnTrack::MatchType t;
+    if (OutboundConnTrack::lookup_match_type(src, t)) {
+      *static_cast<OutboundConnTrack::MatchType *>(data) = t;
+    }
+  }};
+
+// Unfortunately these are not used in RecordsConfig.cc so that must be made consistent by hand.
```
Comment (SolidWallOfCode) [inline proxy/http/HttpConnectionCount.cc]: Not all I'd like, but at least the name is only in 2 places now, reducing the annoying typo glitches.

File: proxy/http/HttpServerSession.h [hunk 15]
```diff
@@ -66,30 +66,23 @@ enum {
...
-      sharing_pool(TS_SERVER_SESSION_SHARING_POOL_GLOBAL),
-      enable_origin_connection_limiting(false),
-      connection_count(nullptr),
-      read_buffer(nullptr),
-      server_vc(nullptr),
-      magic(HTTP_SS_MAGIC_DEAD),
-      buf_reader(nullptr)
-  {
-  }
+  HttpServerSession() : super_type(nullptr) {}
```
Comment (SolidWallOfCode) [inline proxy/http/HttpServerSession.h]: As long as I was in this file, I cleaned up the member initialization.

File: proxy/http/HttpTransact.h (hunk 3 above)
Comment (SolidWallOfCode) [reply proxy/http/HttpTransact.h]: I think that will end up worse, because of the heavy interaction with other parts of the `HttpSM`.  `http_do_server_open` is the only place with non-trivial code and that logic touches a lot of `HttpSM` internals. I'd rather have `HttpSM` do that, than have a subclass reach back.

File: src/traffic_server/InkAPI.cc [hunk 16]
```diff
@@ -7801,367 +7801,412 @@ TSSkipRemappingSet(TSHttpTxn txnp, int flag)
 
 template <typename T>
 inline void *
-_memberp_to_generic(T *ptr, std::type_info const *&type)
-{
-  type = &typeid(T);
+_memberp_to_generic(T *ptr, MgmtConverter const *&conv)
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPI.cc]: Rather than just type information, return a converter to manage the interaction. To avoid excess code and updates, this does the type information locally and if it's one of the few types that were checked, provide a generic identity converter.

File: src/traffic_server/InkAPITest.cc [hunk 17]
```diff
@@ -8572,126 +8572,127 @@ EXCLUSIVE_REGRESSION_TEST(SDK_API_TSHttpConnectServerIntercept)(RegressionTest *
...
-                                                             "proxy.config.http.parent_proxy.fail_threshold",
-                                                             "proxy.config.http.parent_proxy.retry_time",
-                                                             "proxy.config.http.parent_proxy.per_parent_connect_attempts",
-                                                             "proxy.config.http.parent_proxy.connect_attempts_timeout",
-                                                             "proxy.config.http.normalize_ae",
-                                                             "proxy.config.http.insert_forwarded",
-                                                             "proxy.config.http.allow_multi_range",
-                                                             "proxy.config.http.request_buffer_enabled",
-                                                             "proxy.config.http.allow_half_open"};
+std::array<std::string_view, TS_CONFIG_LAST_ENTRY> SDK_Overridable_Configs = {
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPITest.cc]: Here's a place where `std::array` is actually better - the requirement the array *must* be exactly the size of the enum can be enforced statically. This means if some one adds an override and forgets to update this array, it won't even compile.

File: src/traffic_server/InkAPITest.cc [hunk 18]
```diff
@@ -8707,10 +8708,22 @@ REGRESSION_TEST(SDK_API_OVERRIDABLE_CONFIGS)(RegressionTest *test, int /* atype
   s->init();
 
   *pstatus = REGRESSION_TEST_INPROGRESS;
-  for (int i = TS_CONFIG_NULL + 1; i < TS_CONFIG_LAST_ENTRY; ++i) {
-    conf = SDK_Overridable_Configs[i];
+  for (int i = 0; i < static_cast<int>(SDK_Overridable_Configs.size()); ++i) {
+    std::string_view conf{SDK_Overridable_Configs[i]};
 
-    if (TS_SUCCESS == TSHttpTxnConfigFind(conf, -1, &key, &type)) {
+    if (TS_SUCCESS == TSHttpTxnConfigFind(conf.data(), -1, &key, &type)) {
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPITest.cc]: Might as well check both the unsized and sized variants.

File: mgmt/MgmtDefs.h [hunk 19]
```diff
@@ -61,6 +63,35 @@ typedef enum {
...
+  // MgmtFloat conversions.
+  std::function<MgmtFloat(void *)> get_float{nullptr};
+  std::function<void(void *, MgmtFloat)> set_float{nullptr};
+
+  // MgmtString conversions.
+  // This is a bit different because it takes std::string_view instead of MgmtString but that's
+  // worth the difference.
+  std::function<std::string_view(void *)> get_string{nullptr};
+  std::function<void(void *, std::string_view)> set_string{nullptr};
+};
```
Comment (ywkaras) [inline mgmt/MgmtDefs.h]: Why not just: https://godbolt.org/g/qtR5w7 ?

Do you need to wear a drool bucket every time you think of std::function?

File: doc/admin-guide/files/records.config.en.rst [hunk 20]
```diff
@@ -1354,28 +1354,84 @@ Origin Server Connect Attempts
...
+
+   port
+      Group by IP address and port. Each distinct IP address and port pair is a group.
+
+   host
+      Group by host name. The host name is the post remap FQDN used to resolve the upstream
+      address.
+
+   both
+      Group by IP address, port, and host name. Each distinct combination is a group.
```
Comment (ywkaras) [inline doc/admin-guide/files/records.config.en.rst]: So a host name can sometimes resolve to multiple IP addresses, presumably for some load-balancing purpose?  Otherwise, you'd only want to look at the FQDN / port combination.

File: proxy/http/HttpConnectionCount.cc [hunk 21]
```diff
@@ -21,51 +21,316 @@
...
+int
+Config_Update_Conntrack_Max(const char *name, RecDataT dtype, RecData data, void *cookie)
+{
+  auto config = static_cast<OutboundConnTrack::TxnConfig *>(cookie);
+
+  if (RECD_INT == dtype) {
+    config->max = data.rec_int;
+  }
+  return REC_ERR_OKAY;
+}
```
Comment (ywkaras) [inline proxy/http/HttpConnectionCount.cc]: Or, for more limited cut-n-paste action, https://godbolt.org/g/p5qEF3 .

File: proxy/http/HttpConnectionCount.cc [hunk 22]
```diff
@@ -21,51 +21,316 @@
...
+std::time_t
+OutboundConnTrack::Group::get_last_alert_epoch_time() const
+{
+  return Clock::to_time_t(TimePoint{TimePoint::duration{Ticker{_last_alert}}});
+}
+
+void
+OutboundConnTrack::get(std::vector<Group const *> &groups)
+{
+  ink_scoped_mutex_lock lock(_imp._mutex); // TABLE LOCK
```
Comment (ywkaras) [inline proxy/http/HttpConnectionCount.cc]: std::lock_guard ?

File: mgmt/MgmtDefs.h (hunk 19 above)
Comment (SolidWallOfCode) [reply mgmt/MgmtDefs.h]: Your code doesn't compile if any of the lambdas do a capture.

File: doc/admin-guide/files/records.config.en.rst (hunk 20 above)
Comment (SolidWallOfCode) [reply doc/admin-guide/files/records.config.en.rst]: Yes, a single FQDN can resolve to multiple IP addresses.

File: mgmt/MgmtDefs.h (hunk 19 above)
Comment (ywkaras) [reply mgmt/MgmtDefs.h]: I think captures are only useful for local variables, and I think the instances of this class would never be a local variable, thus no need / point in captures.

File: proxy/http/HttpConnectionCount.cc (hunk 21 above)
Comment (SolidWallOfCode) [reply proxy/http/HttpConnectionCount.cc]: I don't think that will handle the time based ones - you can't just assign an integer to a `std::chrono` instance. There are only two bare assigns, not really worth a template for that.

Comment (ywkaras) [reply proxy/http/HttpConnectionCount.cc]: Well then just write all the needed overloads for Config_Update_Setter()  https://godbolt.org/g/xRW3ht .  Do I have to do every little thing for you or what?

File: proxy/http/HttpConnectionCount.cc [hunk 23]
```diff
@@ -21,51 +21,318 @@
...
+  extent += trailer.size();
+
+  text.resize(extent);
+  ts::FixedBufferWriter w(const_cast<char *>(text.data()), text.size());
+  w.print(header_fmt, groups.size());
+  for (auto g : groups) {
+    printer(w, g);
+  }
+  if (groups.size() > 0 && w.remaining()) {
+    w.auxBuffer()[-2] = ' '; // convert trailing comma to space.
```
Comment (ywkaras) [inline proxy/http/HttpConnectionCount.cc]: Really?  Really?  This is why we can't have nice abstractions.

Comment (SolidWallOfCode) [reply proxy/http/HttpConnectionCount.cc]: Yeah, it's a bit ugly. I'm trying to decide how to do this better. But the buffer is known to be long enough so this works on a `LocalBufferWriter`. I'm open to suggestions.

File: proxy/http/HttpConnectionCount.cc (hunk 21 above)
Comment (SolidWallOfCode) [reply proxy/http/HttpConnectionCount.cc]: I'm just not sure the benefit is worth the effort.

File: mgmt/MgmtDefs.h (hunk 19 above)
Comment (SolidWallOfCode) [reply mgmt/MgmtDefs.h]: It's not performant code so I'd rather have the flexibility. There are already layers of virtual dispatches to get to this point, it's not going to make a detectable difference. You should get over your allergic reaction to `std::function`.

I'm also not confident local variables won't be needed - I can easily see wanting to embed references to other objects needed to do the conversions, such as name tables and such.

File: proxy/http/HttpConnectionCount.h [hunk 24]
```diff
@@ -21,237 +21,382 @@
...
+#include <ts/Diags.h>
+#include <ts/CryptoHash.h>
+#include <ts/BufferWriterForward.h>
+#include <MgmtDefs.h>
 #include "HttpProxyAPIEnums.h"
 #include "Show.h"
-#include <sstream>
 
-#pragma once
+constexpr std::string_view operator"" _sv(const char *s, size_t n)
```
Comment (SolidWallOfCode) [inline proxy/http/HttpConnectionCount.h]: This is annoying - for some reason using this literal constructor enables a `std::array<std::string_view>` to be `constexpr` when using `sv` doesn't. That seems seriously broken to me, but there it is.

**Comment by SolidWallOfCode** [comment]
I think is the final form - all comments have been addressed. I tried using templates for the config update functions but looking at it after, it seemed more baroque than the hand rolled version.

File: mgmt/MgmtDefs.h (hunk 19 above)
Comment (ywkaras) [reply mgmt/MgmtDefs.h]: Captures are only necessary to embed references to local variables I think.

File: proxy/http/HttpConfig.cc (hunk 10 above)
Comment (ywkaras) [reply proxy/http/HttpConfig.cc]: General observation:  why are you putting comments like this in the PR rather than the code itself?

File: proxy/http/HttpConnectionCount.cc (hunk 21 above)
Comment (ywkaras) [reply proxy/http/HttpConnectionCount.cc]: http://www.worldoflongmire.com/features/apes/planet5/PAPERWAD.JPG

File: proxy/http/HttpConnectionCount.cc (hunk 23 above)
Comment (ywkaras) [reply proxy/http/HttpConnectionCount.cc]: Aren't you missing a "fill()" call here anyway?  Can't you just make size of the fill() one less to not include the dangling comma?

File: proxy/http/HttpConfig.cc (hunk 10 above)
Comment (SolidWallOfCode) [reply proxy/http/HttpConfig.cc]: Because the comment will just be noise once the PR is committed.

File: mgmt/MgmtDefs.h (hunk 19 above)
Comment (SolidWallOfCode) [reply mgmt/MgmtDefs.h]: `std::function` is cleaner and simpler and therefore should be used by default, unless there is a good reason to write more complicated code, such as for tight performance constraints.

File: proxy/http/HttpConnectionCount.cc (hunk 23 above)
Comment (SolidWallOfCode) [reply proxy/http/HttpConnectionCount.cc]: No. There is no missing call to `fill`. I wonder if it might work to `clip(2)` and then paste the terminator in. Have to try that. The `std::string` has been resized to be the exact expected output length at that point.

File: mgmt/MgmtDefs.h (hunk 19 above)
Comment (ywkaras) [reply mgmt/MgmtDefs.h]: I don't see how using std::function is simpler than just a function pointer when that will do.

File: proxy/http/HttpConnectionCount.cc (hunk 23 above)
Comment (ywkaras) [reply proxy/http/HttpConnectionCount.cc]: Hmmm not seeing that.

File: proxy/http/HttpConfig.h (hunk 12 above)
Comment (zwoop) [reply proxy/http/HttpConfig.h]: 👍

File: src/traffic_server/InkAPITest.cc (hunk 17 above)
Comment (zwoop) [reply src/traffic_server/InkAPITest.cc]: 👍

File: mgmt/MgmtDefs.h (hunk 19 above)
Comment (SolidWallOfCode) [reply mgmt/MgmtDefs.h]: As noted, I am not confident a simple function pointer will do. I ended up doing that inside some of the buffer writer formatting and now it's a nasty problem preventing me from doing things I want to do. I think micro-optimizing, especially in cases like this where performance isn't critical, is usually a bad choice.
